File: src/main/clojure/euler2.clj
(ns euler2)

; given an adjacent pair of numbers in the fibonacci sequence, calculate the next pair
(defn next-fib-pair [[a b]]
  [b (+ a b)])

; Calculate the sum of all even fibonacci numbers less than the given number
(defn fib-sum-even [max-fib]
  (loop [fib-pair [0 1] acc 0]
    (let [fibn (last fib-pair)]
      (if (> fibn max-fib)
        acc
        (if (zero? (mod fibn 2))
          (recur (next-fib-pair fib-pair) (+ acc fibn))
          (recur (next-fib-pair fib-pair) acc)
)))))

; Problem 2: find the sum of all even fibonacci numbers less than 4 million
(println
  (fib-sum-even 4000000)
  )

File: src/main/ruby/euler2.rb
def nextFib(curFib)
  [curFib[1], curFib[0] + curFib[1]]
end

curFib = [1, 2]
sum = 2
while curFib[1] < 4000000
  curFib = nextFib(curFib)
  sum += curFib[1] if curFib[1].even?
end

puts "Sum of even Fibonacci numbers less than 4 million: #{sum}"
